Offset the line to fit the glyph with negative side bearing

w0rm opened the pull request:
Implements the logic that we discussed in #166.

This is how it works for MogeeFont (with justified text):
<img width="711" alt="Screenshot 2024-04-28 at 17 27 25" src="https://github.com/embedded-graphics/embedded-text/assets/43472/033ebdee-f4fa-4b97-8443-e978e0dbc606">

bugadani commented on src/rendering/line_iter.rs:
```diff
@@ -321,6 +324,15 @@ where
+                    if self.empty {
+                        // If this is the first word on the line, offset the line by
+                        // the word's left negative boundary to make sure it is not clipped.
+                        let offset = handler.measure_left_offset(w);
```
Can we somehow avoid double-measuring here? I'd probably prefer modifying `measure` to return the offset along with the width.

w0rm replied:
Done! Realised that having a more specialised function was nicer than changing the existing `measure` function, to avoid touching a lot of places where it is used.

bugadani commented:
Thanks for working on this, this is an even smaller diff than I expected :)

w0rm commented:
@bugadani I started with a much more complex solution and then realised that simply injecting unaccounted whitespace did the trick! 

Will update the existing measure function to prevent from double measuring.

bugadani commented:
Thank you, this is awesome. Two more things and I think we can ship this:

 - could you please add a rendering test like https://github.com/embedded-graphics/embedded-text/blob/f49d5243ece537b1cd8ed5bb83d7b8448c223515/src/rendering/mod.rs#L337 to make it obvious this PR does what you intended?
 - could you please add this to the changelog? :)

w0rm commented:
@bugadani rendering testing turned out to be quite involved, because I needed to implement a test font. The font I am working on is not released yet, and it might not be a good idea to depend on it.

w0rm commented:
@bugadani please review but don’t merge this just yet. I want to run a few more tests next week!

bugadani commented:
Apologies, I didn't intend the test to turn this hairy, not sure what went through my head with that. However, thanks for adding it!

If you'd like to prevent me from merging, just feel free to mark the PR as a draft (although... request noted even though the green button is tempting :) ), you can find the option top right, under the "Reviewers" box.

w0rm commented:
@bugadani unfortunately I found another issue.. this time it is a panic when splitting long words in [`longest_fitting_substr`](https://github.com/embedded-graphics/embedded-text/blob/master/src/rendering/line_iter.rs#L134). 

A pixel width of a str is not equal to a sum of widths of individual characters. In my font, the space between a pair of characters depends on the characters themselves (this is called kerning),  and also sometimes consecutive characters may be collapsed into a ligature. 

In the test font that I added, the spacing between the characters is 1px that is not accounted into the char width. This makes the actual length of the string longer than the sum of widths of individual characters. I added a test that panics.

I can replace this with a function that iterates over the character indices and measures the str width until that, but it would certainly be slower than the current implementation.

bugadani commented:
> A pixel width of a str is not equal to a sum of widths of individual characters.

Yes, you are bringin in a whole new set of assumptions. e-t was built assuming fonts are monospaced, so in several places `width = length * character_width`.

At this point I can accept compromises in terms of speed, as some of the "optimisations" present were made without any instrumentation. We can tune things later if speed becomes an issue, although I'm quite sure nobody build an e-book reader with e-t :)

w0rm commented:
@bugadani I've committed a possible fix for this issue and added tests for TestFont and MonoFont.

bugadani commented:
I do not but that is mainly because I haven't read this code in a long time.

w0rm commented:
I looked up all the calls to `measure`, and the only remaining place that I noticed is a place that assumes that the break character starts when the word ends. 

```rust
Some(Token::Break(w)) => return Some(width + handler.measure(w)),
```

I am not relying on break characters in my text, we can tackle this issue separately when this becomes a problem. It renders a hyphen without a space, but that's not as bad as a panic.

<img width="558" alt="Screenshot 2024-04-29 at 20 13 17" src="https://github.com/embedded-graphics/embedded-text/assets/43472/3dad0698-77c7-4196-a7fb-0b1623de60f2">

bugadani commented:
Thank you very much for working on this!
